File: algorithms/sorting/heap-sort/HeapSort.java
import java.util.Arrays;

public final class HeapSort {
    private HeapSort() {
    }

    public static int[] heapSort(int[] numbers) {
        int[] array = Arrays.copyOf(numbers, numbers.length);

        for (int root = array.length / 2 - 1; root >= 0; root--) {
            siftDown(array, root, array.length);
        }

        for (int end = array.length - 1; end > 0; end--) {
            int temporary = array[0];
            array[0] = array[end];
            array[end] = temporary;
            siftDown(array, 0, end);
        }

        return array;
    }

    private static void siftDown(int[] array, int root, int size) {
        while (2 * root + 1 < size) {
            int left = 2 * root + 1;
            int right = left + 1;
            int largest = left;

            if (right < size && array[right] > array[left]) largest = right;
            if (array[root] >= array[largest]) return;

            int temporary = array[root];
            array[root] = array[largest];
            array[largest] = temporary;
            root = largest;
        }
    }
}

File: algorithms/sorting/heap-sort/heap_sort.py
def heap_sort(numbers):
    array = numbers.copy()

    for root in range(len(array) // 2 - 1, -1, -1):
        sift_down(array, root, len(array))

    for end in range(len(array) - 1, 0, -1):
        array[0], array[end] = array[end], array[0]
        sift_down(array, 0, end)

    return array


def sift_down(array, root, size):
    while 2 * root + 1 < size:
        left = 2 * root + 1
        right = left + 1
        largest = left

        if right < size and array[right] > array[left]:
            largest = right
        if array[root] >= array[largest]:
            return

        array[root], array[largest] = array[largest], array[root]
        root = largest
